xcdat/averager.py

```diff
@@ -299,52 +338,37 @@ def _averager(xvariable, weights, axis):
     ----------
     xvariable : xr.DataArray
         xr.DataArray produced using ``open_variable``
-
     weights : xr.DataArray
         DataArray containing the desired weights for the variable array
-
-    axis : list
-        List of string values denoting the dims that should be averaged over
-        (e.g., axis=['lat', 'lon'])
+    axis : SupportedAxes
+        List of string values denoting the dims that should be averaged over.
 
     Returns
     -------
     xr.DataArray
-        Dataarray that has been reduced via a weighted average.
+        DataArray that has been reduced via a weighted average.
 
     Notes
     -----
     Needs to be evaluated for speed and compared to native xarray and CDAT
     functionality.
-
     """
-    # get masked data
-    masked = np.isnan(xvariable)
-    # ensure weights includes all of the same axes that are in the
-    # variable array
-    for i, dim in enumerate(xvariable.dims):
+    # Ensure weights includes all of the same axes that are in the
+    # variable array wtih identical size.
+    for dim, size in xvariable.sizes.items():
         if dim not in weights.dims:
-            weights = weights.expand_dims(dim)
-    # tranpose to ensure axis ordering is the same
+            weights = weights.expand_dims({dim: size})
+    # Tranpose to ensure axis ordering is the same and fill null values.
     weights = weights.transpose(*xvariable.dims)
-    # expand array so that the variable and weights arrays have identical shape
-    # REVIEW: It may be useful to create a function (e.g., genutil.grower)
-    #         that does this operation.
-    ntile = []
-    for i, dim in enumerate(weights.dims):
-        if len(weights[dim]) != len(xvariable[dim]):
-            ntile.append(len(xvariable[dim]))
-        else:
-            ntile.append(1)
-    weights = np.tile(weights, ntile)
-    # zero out missing data
-    weights[masked] = 0.
+    weights = weights.fillna(0)
+
+    # TODO: Below chunk of code should be extracted into a seperate function.
     # get averaging axis indices
```

Comment:
I was able to refactor this entire chunk of code which is supposed to replicate `genutil.grower` using the initial `expand_dims` call. The `expand_dims` `dim` arg allows you to specify the dimension and size of the dimension to grow the dataarray.

https://xarray.pydata.org/en/stable/generated/xarray.DataArray.expand_dims.html